fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,258 @@ impl ExecutorInner {
                 }
-            };
-            let (completed_result, _index, remaining_transitions) = select! {
-                res = select_all(transitions) => res,
-                () = new_state_added => {
-                    debug!("New state added, re-starting state transitions");
-                    return Ok(());
-                }
-            };
-            transitions = remaining_transitions;
-            let (transition_outcome, state, transition_fn, meta) = completed_result;
-            let operation_id = state.operation_id();
-            // info level because spans only log when any event happens within a span.
-            let span = tracing::info_span!("state_machine_transition", %operation_id);
-            async {
-                info!("Executing state transition");
-                debug!(
-                    ?state,
-                    transition_outcome = ?AbbreviateJson(&transition_outcome),
-                );
-
-                let active_or_inactive_state = self
-                    .db
-                    .autocommit(
-                        |dbtx, _| {
-                            let state = state.clone();
-                            let transition_fn = transition_fn.clone();
-                            let transition_outcome = transition_outcome.clone();
-                            Box::pin(async move {
-                                let new_state = transition_fn(
-                                    &mut ClientSMDatabaseTransaction::new(
-                                        &mut dbtx.to_ref(),
-                                        state.module_instance_id(),
-                                    ),
-                                    transition_outcome,
-                                    state.clone(),
-                                )
-                                .await;
-                                dbtx.remove_entry(&ActiveStateKey::from_state(state.clone()))
-                                    .await;
-                                dbtx.insert_entry(
-                                    &InactiveStateKey::from_state(state.clone()),
-                                    &meta.into_inactive(),
+                ExecutorLoopEvent::Triggered(TransitionForActiveState {
+                    outcome,
+                    state,
+                    meta,
+                    transition_fn,
+                }) => {
+                    let span = tracing::info_span!(
+                        "state_machine_transition",
+                        operation_id = %state.operation_id()
+                    );
+                    // Perform the transition as another future, so transitions can happen in
+                    // parallel.
+                    // Database write conflicts might be happening quite often here,
+                    // but transaction functions are supposed to be idempotent anyway,
+                    // so it seems like a good stress-test in the worst case.
```

Comment:
Depending on how heavy operations are run here that could be quite inefficient, imagine multiple SMs trying to create an e-cash output for the same index, all hashing to the BLS curve, which takes a few ms. Let's see how it goes in practice.